### PR TITLE
Public avatar changes/clean-up

### LINKED ISSUE #572: Ability to remove avatar layers when viewing bundles
When creating an avatar and viewing unlocked bundles there is no way to remove an item without navigating away from the item and going to its individual layer.

### LINKED ISSUE #595: Avatar selection internationalization
Ensure all avatar selection code is i18n and that l10n exists for Spanish. Check the "No <item>" button as well.

### PR DESCRIPTION
**Pull request by madison-boman**
Cleanup of frontend code, makes bundle items removable and improves i10n/i18n

Fix #572 Ability to remove avatar layers when viewing bundles
Fix #595 Avatar selection internationalization

### CODE REVIEW

**Comment by madison-boman** [comment]
- 📇 Includes database changes

File: src/GRA.Domain.Service/AvatarService.cs [hunk 1]
```diff
@@ -500,6 +516,7 @@ public async Task UpdateUserAvatarAsync(ICollection<AvatarLayer> selectionLayers
         public async Task<AvatarBundle> GetBundleByIdAsync(int id, bool includeDeleted = false)
         {
             var bundle = await _avatarBundleRepository.GetByIdAsync(id, includeDeleted);
+            bundle.AvatarItems = await _avatarItemRepository.GetBundleItemsAsync(id);
```
Comment (iafb) [inline src/GRA.Domain.Service/AvatarService.cs:519]: I think AvatarItems are already being set in the bundle through AutoMapper and we don't need to do an additional call to set them.

File: src/GRA/Annotations.cs [hunk 2]
```diff
@@ -77,6 +78,7 @@ public static class Interface
             public const string AndTheir = "and their {0}";
             public const string Apply = "Apply";
             public const string AvatarSaved = "Avatar saved.";
+            public const string Bundle = "Bundle";
```
Comment (iafb) [inline src/GRA/Annotations.cs:81]: This annotation is missing from the .resx

File: src/GRA.Domain.Service/AvatarService.cs [hunk 3]
```diff
@@ -431,34 +444,37 @@ public async Task<ICollection<AvatarElement>> GetUserAvatarAsync()
...
 
         public async Task UpdateUserAvatarAsync(ICollection<AvatarLayer> selectionLayers)
         {
             var activeUserId = GetActiveUserId();
-            var layers = await _avatarLayerRepository.GetAllAsync(GetCurrentSiteId());
+            var currentCultureName = _userContextProvider.GetCurrentCulture()?.Name;
+            var currentLanguageId = currentCultureName != null ?
+                await _languageService.GetLanguageIdAsync(currentCultureName) :
+                await _languageService.GetDefaultLanguageIdAsync();
+            VerifyManagementPermission();
```
Comment (iafb) [inline src/GRA.Domain.Service/AvatarService.cs:476]: VerifyManagementPermission shouldn't be called here

File: src/GRA.Controllers/AvatarController.cs [hunk 4]
```diff
@@ -97,13 +87,11 @@ public async Task<JsonResult> SaveAvatar(string selectionJson)
         }
 
         [HttpPost]
-        public async Task<JsonResult> UpdateUserLogs(string selectionJson)
+        public async Task<JsonResult> UpdateBundleHasBeenViewed(int bundleId)
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/GRA.Controllers/AvatarController.cs:90]: Resolved in commit 6354afbeb309b82300906387d493f51f1ba975dd

File: src/GRA.Controllers/AvatarController.cs [hunk 5]
```diff
@@ -210,6 +198,54 @@ public async Task<IActionResult> Share(string selectionJson)
...
 
+        public async Task<IActionResult> GetLayersItems(
+            string type, int layerId, int selectedItemId, int bundleId, int[] selectedItemIds)
+        {
+            bool user = !AuthUser.Identity.IsAuthenticated;
+            if (user)
+            {
+                ShowAlertDanger(_sharedLocalizer[Annotations.Validate.SessionExpired].ToString());
+                return NotFound();
+            }
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/GRA.Controllers/AvatarController.cs]: Resolved in commit 168fc2e0e54709d64bf4cf48f41548ffb39b4c56

File: src/GRA.Data/Model/AvatarLayerText.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        public AvatarLayer AvatarLayer{get;set;}
+
+        [Required]
+        public int LanguageId { get; set; }
+        public Language Language { get; set; }
+
+        [Required]
+        [MaxLength(255)]
+        public string Name { get; set; }
+        public string RemoveLabel { get; set; }
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/GRA.Data/Model/AvatarLayerText.cs:20]: Resolved in commit 6edbabd33dcd75f3f1470675af954edb686efd62

File: src/GRA.Controllers/ViewModel/Avatar/AvatarViewModel.cs [hunk 7]
```diff
@@ -7,11 +7,20 @@ public class AvatarViewModel
...
         public string AvatarBundlesJson { get; set; }
         public bool NewAvatar { get; set; }
+        public bool HasUnviewedBundles { get; set; }
+        public ICollection<AvatarItem> LayerItems{get; set;}
+        public ICollection<AvatarColor> LayerColors { get; set; }
+        public int SelectedItemId { get; set; }
+        public int[] SelectedItemIds { get; set; }
+        public string ItemPath { get; set; }
+        public int LayerId { get; set; }
+        public bool UseEnglish { get; set; }
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/GRA.Controllers/ViewModel/Avatar/AvatarViewModel.cs]: Resolved in Commit b64da02bab5edca4a34ae87f500cb9b5ba51edb1

File: src/GRA/Annotations.cs (hunk 2 above)
Comment (madison-boman) [reply src/GRA/Annotations.cs:81]: Resolved in commit b2f1144d79586ad0f605fc9354432a8a201a8f5c

File: src/GRA.Controllers/SignInController.cs [hunk 8]
```diff
@@ -48,7 +48,7 @@ public async Task<IActionResult> Index(string ReturnUrl = null)
                 sendTo = TempData[TempDataKey.ReturnUrl].ToString();
                 TempData.Remove(TempDataKey.ReturnUrl);
             }
-
+            Response.StatusCode = 404;
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/GRA.Controllers/SignInController.cs]: Fixed in commit 43342d8bdb798e95be7285fc34d098fa3f0324e5

File: src/GRA.Data/Repository/AvatarLayerRepository.cs [hunk 9]
```diff
@@ -18,24 +18,50 @@ public class AvatarLayerRepository
...
+                foreach (var layer in layers.ToList())
+                {
+                    var layerText = _context.AvatarLayerText
+                    .AsNoTracking()
+                    .Where(_ => _.AvatarLayerId == layer.Id && _.LanguageId == languageId)
+                    .FirstOrDefault();
+                    layer.Name = layerText.Name;
+                    layer.RemoveLabel = layerText.RemoveLabel;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/GRA.Data/Repository/AvatarLayerRepository.cs]: Fixed in commit a20123bde9343e692f3adfa31bfb53134df12183

File: src/GRA.Data/Repository/AvatarLayerRepository.cs [hunk 10]
```diff
@@ -18,24 +18,50 @@ public class AvatarLayerRepository
...
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/GRA.Data/Repository/AvatarLayerRepository.cs]: Fixed in commit a20123bde9343e692f3adfa31bfb53134df12183

File: src/GRA.Domain.Service/AvatarService.cs (hunk 3 above)
Comment (madison-boman) [reply src/GRA.Domain.Service/AvatarService.cs]: Fixed in commit 1e29452a573b621d3ecb1dfc99d6cf300a36f0b7

File: src/GRA.Domain.Service/AvatarService.cs (hunk 1 above)
Comment (madison-boman) [reply src/GRA.Domain.Service/AvatarService.cs]: Fixed in commit 3fd2916ba9d6500015a137fc3b7ceebcca5cd5f5

File: src/GRA.Domain.Service/AvatarService.cs [hunk 11]
```diff
@@ -178,7 +190,8 @@ public async Task<AvatarLayer> UpdateLayerAsync(AvatarLayer layer)
 
         public async Task<ICollection<AvatarItem>> GetItemsByLayerAsync(int layerId)
         {
-            return await _avatarItemRepository.GetByLayerAsync(layerId);
+            var userId = GetClaimId(ClaimType.UserId);
+            return await _avatarItemRepository.GetUserItemsByLayerAsync(userId,layerId);
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/GRA.Domain.Service/AvatarService.cs]: Fixed in commit 1141db329804052035b3f78db7f6262b171ffe36

File: src/GRA.Controllers/AvatarController.cs [hunk 12]
```diff
@@ -210,6 +198,54 @@ public async Task<IActionResult> Share(string selectionJson)
...
+                return PartialView("_SlickPartial", model);
+            }
+            catch(GraException gex)
+            {
+                _logger.LogError(gex,
+                    "Could not retrieve layer items for layer id {layerId}: {Message}",
+                    layerId,
+                    gex.Message);
+                ShowAlertDanger(_sharedLocalizer[Annotations.Validate.SomethingWentWrong].ToString());
+                return NotFound();
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/GRA.Controllers/AvatarController.cs]: Fixed in 29d78590d1ee3de2bf3e7efe6723d5d9d08fd7e1

**Comment by k7hpn** [comment]
@madison-boman @iafb this is a big PR that's been open for a loooooong time. What do we need to do to get this merged?

**Comment by madison-boman** [comment]
> @madison-boman @iafb this is a big PR that's been open for a loooooong time. What do we need to do to get this merged?

@k7hpn I could condense it like the acessibility PR's to make it easier for review

**Comment by iafb** [comment]
> 
> 
> @madison-boman @iafb this is a big PR that's been open for a loooooong time. What do we need to do to get this merged?

To my knowledge the changes in this PR were good and ready to be merged.

**Comment by k7hpn** [comment]
@madison-boman @iafb If this is really good to go then just bring it up to date with develop and then I'll give it a final review and we'll merge it!

**Comment by madison-boman** [comment]
@k7hpn This is all up to date with develop!

File: src/GRA.Web/Views/Avatar/_SlickPartial.cshtml [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+            </div>
+        }
+    </div>
+}
+@if (Model.Bundle != null)
+{
+    int index;
+    try
+    {
+        index = Model.SelectedItemIds
```
Comment (k7hpn) [inline src/GRA.Web/Views/Avatar/_SlickPartial.cshtml:51]: This logic doesn't really belong in the view, can we move this out to a controller?

**Review by k7hpn — CHANGES_REQUESTED** [review]
One bit of logic that's using Linq in a view, not really the best place for it.

File: src/GRA.Web/Views/Avatar/_SlickPartial.cshtml (hunk 13 above)
Comment (madison-boman) [reply src/GRA.Web/Views/Avatar/_SlickPartial.cshtml]: Just pushed these changes!

File: src/GRA.Web/Views/Avatar/_SlickPartial.cshtml [hunk 14]
```diff
@@ -45,29 +45,10 @@
 }
 @if (Model.Bundle != null)
 {
-    int index;
-    try
-    {
-        index = Model.SelectedItemIds
```
Comment (k7hpn) [inline src/GRA.Web/Views/Avatar/_SlickPartial.cshtml]: What exactly is this Linq doing? It seems overly complex.

**Pull request merged**